### PR TITLE
added savegame compat code

### PR DESCRIPTION
**Pull request by hessenfarmer**
fixed broken savegame compatibility due to 3 programs productivity change

### CODE REVIEW

**Comment by Noordfrees** [comment]
IMHO having a specialised lookup table is a bit overkill. Just rename one of the programs back to the original name.

**Comment by hessenfarmer** [comment]
Basically you are right however I implemented it the way other savegame issues were solved too. So I believe as it is working this way it would help us with future compatibility issues if necessary. And as this is only used in loading at startup it shouldn't cost much.

**Comment by Noordfrees** [comment]
But if we now add a fifth tribe which has a building with a program named "smelt_iron", this will also be looked up and changed to "smelt_iron_1", which does not exist. So looking up program names always implicitly introduces program name restrictions and program substitution bugs. 
If you want to keep this code, you need to increase the packet version and only call the lookup for savegames with the older packet version.

File: src/map_io/map_buildingdata_packet.cc [hunk 1]
```diff
@@ -681,6 +681,7 @@ void MapBuildingdataPacket::read_productionsite(
 			//  skipped programs
 			uint32_t const gametime = game.get_gametime();
 			for (uint8_t i = fr.unsigned_8(); i; --i) {
+
```
Comment (Noordfrees) [inline src/map_io/map_buildingdata_packet.cc]: Please remove the newline again

**Comment by hessenfarmer** [comment]
> 
> 
> But if we now add a fifth tribe which has a building with a program named "smelt_iron", this will also be looked up and changed to "smelt_iron_1", which does not exist. So looking up program names always implicitly introduces program name restrictions and program substitution bugs.
> If you want to keep this code, you need to increase the packet version and only call the lookup for savegames with the older packet version.

Very valid point so I'll keep the lookup code in (for possible future use, but empty the table while implementing it like you suggested.

**Review by Noordfrees — APPROVED** [review]
Code looking good now :)
untested

**Comment by gunchleoc** [comment]
How about resetting to the start if a program isn't found? The game state won't be perfect, but we won't need to keep a table.

**Comment by hessenfarmer** [comment]
> 
> 
> How about resetting to the start if a program isn't found? The game state won't be perfect, but we won't need to keep a table.

Could be done as well. However this should go in as it fixes the break in savegame compatibility I accidentally introduced with the 3 programs performance improvement.

**Comment by gunchleoc** [comment]
Can you open a new issue before we merge this, so we won't forget?

**Comment by hessenfarmer** [comment]
done see #3671

**Comment by gunchleoc** [comment]
Thanks! Let's squash & merge when the CI is ready.

**Pull request merged**